presto-main/src/main/java/com/facebook/presto/operator/SpatialJoinOperator.java

```diff
@@ -0,0 +1,276 @@
+        this.pagesSpatialIndexFuture = pagesSpatialIndexFactory.createPagesSpatialIndex();
+        this.outputTypes = ImmutableList.copyOf(outputTypes);
+        this.pageBuilder = new PageBuilder(outputTypes);
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public List<Type> getTypes()
+    {
+        return outputTypes;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finished && pagesSpatialIndexFuture.isDone() && !pageBuilder.isFull() && probe == null;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        verify(probe == null);
+        probe = page;
+        probePosition = 0;
+
+        joinAddresses = null;
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        verify(!finished);
+        if (!pageBuilder.isFull() && probe != null) {
+            processProbe();
+        }
+
+        if (pageBuilder.isFull()) {
+            Page page = pageBuilder.build();
+            pageBuilder.reset();
+            return page;
+        }
+
+        if (finishing && probe == null) {
+            Page page = null;
+            if (!pageBuilder.isEmpty()) {
+                page = pageBuilder.build();
+                pageBuilder.reset();
+            }
+            pagesSpatialIndexFactory.probeOperatorFinished();
+            pagesSpatialIndexFuture = null;
+            finished = true;
+            return page;
+        }
+
+        return null;
+    }
+
+    private void processProbe()
+    {
+        verify(probe != null);
+
+        if (pagesSpatialIndex == null) {
+            pagesSpatialIndex = getDone(pagesSpatialIndexFuture).get();
+        }
+
+        DriverYieldSignal yieldSignal = operatorContext.getDriverContext().getYieldSignal();
+        while (probePosition < probe.getPositionCount()) {
+            if (joinAddresses == null) {
+                joinAddresses = pagesSpatialIndex.findJoinAddresses(probePosition, probe, probeGeometryChannel);
+                nextJoinAddressIndex = 0;
+                if (yieldSignal.isSet()) {
+                    return;
+                }
+            }
+
+            while (nextJoinAddressIndex < joinAddresses.length) {
+                if (pageBuilder.isFull()) {
+                    return;
+                }
+
+                long joinAddress = joinAddresses[nextJoinAddressIndex];
+
+                if (pagesSpatialIndex.isJoinAddressEligible(joinAddress, probePosition, probe)) {
+                    pageBuilder.declarePosition();
+                    int outputChannelOffset = 0;
+                    for (int outputIndex : probeOutputChannels) {
+                        Type type = probeTypes.get(outputIndex);
+                        Block block = probe.getBlock(outputIndex);
+                        type.appendTo(block, probePosition, pageBuilder.getBlockBuilder(outputChannelOffset));
```

Comment:
@highker would it be possible to reuse `LookupJoinPageBuilder` here to avoid copying?
Otherwise, we could directly use `DictionaryBlock`s for probe to avoid copying.. This can be as a follow-up.

Reply:
Yes, actually I think we should use `LookupJoinPageBuilder` here. Recall that `appendTo` would be 20% slower given https://github.com/prestodb/presto/pull/9423 has moved `joinProbeCompiler`.